### PR TITLE
Move `upgrade` integration tests in their own files

### PR DESCRIPTION
**Pull request by mateiidavid**
This change hoists all of the upgrade logic into dedicated files.

Upgrade tests are now done programatically (at the expense of repeating a lot of code between `upgrade-edge` and `upgrade-stable`). This (in theory) will remove the dependency on an external test-runner and special arguments for the `TestHelper`. Upgrade tests can now be run through go's runner, provided a Kubernetes cluster has been provisioned: `go test ./test/integration/upgrade-<release-channel>/...`

To implement this change, all of the upgrade code has been moved from `install_test.go`. We also create a few additional helper functions:

* we create a helper function to get us the latest linkerd version from both release channels;
* we create a helper function to download the latest binary for a specific release channel;
* and finally, we create a helper function to exec other processes (so we can use the linkerd cli to install the control plane before an upgrade).

### CODE REVIEW

File: test/integration/upgrade-edge/upgrade_edge_test.go [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+
+func TestVersionPostInstall(t *testing.T) {
+	err := TestHelper.CheckVersion(TestHelper.GetVersion())
+	if err != nil {
+		testutil.AnnotatedFatalf(t, "Version command failed",
+			"Version command failed\n%s", err.Error())
+	}
+}
+
+func testCheckCommand(t *testing.T, stage, expectedVersion, namespace, cliVersionOverride string) {
```
Comment (alpeb) [inline test/integration/upgrade-edge/upgrade_edge_test.go:457]: I didn't see where the `linkerd check` commands are performed, calling this...

File: test/integration/upgrade-edge/upgrade_edge_test.go [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+		}
+
+		return nil
+	})
+	if err != nil {
+		testutil.AnnotatedFatal(t, fmt.Sprintf("'linkerd check' command timed-out (%s)", timeout), err)
+	}
+}
+
+func getCheckOutput(t *testing.T, goldenFile string, namespace string) string {
```
Comment (alpeb) [inline test/integration/upgrade-edge/upgrade_edge_test.go:498]: And so doesn't this one.

File: test/integration/upgrade-edge/upgrade_edge_test.go [hunk 3]
```diff
@@ -0,0 +1,533 @@
...
+func TestUpgradeTestAppWorksBeforeUpgrade(t *testing.T) {
+	ctx := context.Background()
+	testAppNamespace := "upgrade-test"
+
+	// create namespace and install test app
+	if err := TestHelper.CreateDataPlaneNamespaceIfNotExists(ctx, testAppNamespace, map[string]string{}); err != nil {
+		testutil.AnnotatedFatalf(t, "Failed to create namespace", "Failed to create namespace %s: %s", testAppNamespace, err)
+	}
+
+	if _, err := TestHelper.Kubectl("", "apply", "-f", "./testdata/emoji.yaml", "-n", testAppNamespace); err != nil {
```
Comment (alpeb) [inline test/integration/upgrade-edge/upgrade_edge_test.go:150]: Originally, emoji was injected in `bin/_test-helpers.sh` so we check the linkerd upgrade doesn't break the app. Here you can instead just annotate the emojivoto namespace for injection, and then before `TestUpgradeTestAppWorksAfterUpgrade` do a dataplane rollout.

**Review by alpeb — COMMENTED** [review]
> Upgrade tests can now be run through go's runner, provided a Kubernetes cluster has been provisioned: go test ./test/integration/upgrade-<release-channel>/...

:100: :medal_military: :trophy:

File: bin/_test-helpers.sh [hunk 4]
```diff
@@ -433,47 +433,20 @@ install_version() {
...
-
-  local upgrade_version
-  upgrade_version=$(latest_release_channel "$release_channel")
-
-  if [ -z "$upgrade_version" ]; then
-    echo 'error getting upgrade_version'
-    exit 1
-  fi
-
-  install_version "$install_url" "$upgrade_version"
```
[the remark this answers is not in the document]
Comment (mateiidavid) [reply bin/_test-helpers.sh:448]: Oh nice! I wanted to make as fewer changes in the bash script as possible so I can avoid messing up the CI. I didn't experiment a lot there :D Great spot.

File: test/integration/upgrade-edge/upgrade_edge_test.go (hunk 3 above)
Comment (mateiidavid) [reply test/integration/upgrade-edge/upgrade_edge_test.go:142]: When creating the namespace we now add the inject annotation. Before we upgrade, we roll out the pods so they get re-injected, block until ready, and then go ahead with the rest of the test 👍🏻  great suggestion

**Review by adleong — APPROVED** [review]

**Review by alpeb — COMMENTED** [review]
I see you removed the functions for doing `linkerd check`, but the golden files are still there. OTOH, maybe it's enough to just check `linkerd check` returns 0, without having to check a golden file?

**Comment by mateiidavid** [comment]
> I see you removed the functions for doing `linkerd check`, but the golden files are still there. OTOH, maybe it's enough to just check `linkerd check` returns 0, without having to check a golden file?

Added a check for the proxy. I think even if the checks fail it'll return 0; I opted to instead just check the proxy output is the same as what we are getting.

**Review by alpeb — APPROVED** [review]
Thanks @mateiidavid ! :ship:

**Pull request merged**